packages/nuxt3/src/app/composables/state.ts

```diff
@@ -1,14 +1,26 @@
+/* eslint-disable no-redeclare */
 import { toRef } from 'vue'
 import type { Ref } from 'vue'
 import { useNuxtApp } from '#app'
 
+const stateSymbol = Symbol('nuxt-states')
+
+export interface NuxtStates { }
+export type StateKey<T> = string & { [stateSymbol]: T }
+
+/** Return a typed string that can be used as a key for `useState`. */
+export const stateKey = <T> (k: string) => k as StateKey<T>
+
 /**
  * Create a global reactive ref that will be hydrated but not shared across ssr requests
  *
  * @param key a unique key ensuring that data fetching can be properly de-duplicated across requests
  * @param init a function that provides initial value for the state when it's not initiated
  */
-export const useState = <T> (key: string, init?: (() => T)): Ref<T> => {
+
+export function useState<T = unknown, K extends string = string, S = K extends keyof NuxtStates ? NuxtStates[K] : T> (key: K | StateKey<T>): Ref<S | undefined>
+export function useState<T = unknown, K extends string = string, S = K extends keyof NuxtStates ? NuxtStates[K] : T> (key: K | StateKey<T>, init: (() => K extends keyof NuxtStates ? NuxtStates[K] : T)): Ref<S>
+export function useState (key: string, init?: (() => any)): Ref<any> {
```

Comment:
Added NuxtStates key hint, replaced duplicate type for init method return value, removed redundant any type
```suggestion
export function useState<T = unknown, K extends string | keyof NuxtStates = string, S = K extends keyof NuxtStates ? NuxtStates[K] : T> (key: K | StateKey<T>): Ref<S | undefined>
export function useState<T = unknown, K extends string | keyof NuxtStates = string, S = K extends keyof NuxtStates ? NuxtStates[K] : T> (key: K | StateKey<T>, init: (() => S)): Ref<S>
export function useState (key: string, init?: (() => any)): Ref {
```

Reply:
the duplicate type for init method return value is necessary for TS to complain if you return the wrong type from the init - otherwise it silently reshapes the type.

I also don't think adding `| keyof NuxtStates` gives us any key hint, but please let me know if it's different for you.

Reply:
didn't realise about the duplicate 👍 

https://user-images.githubusercontent.com/41805560/141422385-10fb50cc-2592-4db7-afa6-68638b0ccd5b.mov

That's in jetbrains IDE